### PR TITLE
[SGD] Variable worker CPU requirements

### PR DESCRIPTION
**Pull request by ConeyLiu**
## Why are these changes needed?

Support set the CPU resource requirements for the training worker. This is useful for CPU only distributed model training to avoid CPU resource competition. And we also need extra processes to do speed up data loading, so the CPU resources should be specified.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Comment by ConeyLiu** [comment]
Hi @richardliaw, could you help to review on this? Thanks a lot. And also, I have a question whether should we support some resource init function such as the following?
```python
def resource_requirement(rank_id):
    #  specify the resource requirement for the given rank worker
    return resources
```
I think this is helpful because of user can set the specified resource requirement for each training worker.

File: python/ray/util/sgd/torch/distributed_torch_runner.py [hunk 1]
```diff
@@ -220,28 +228,23 @@ def reserve_cuda_device(retries=20):
...
             logger.debug("Device %s not in list of visible devices %s",
-                         reserved_device, cuda_device_set)
-            unused_actors.append(_dummy_actor)
+                         reserved_gpu_device, cuda_device_set)
+            _dummy_actor.__ray_terminate__.remote()
             _dummy_actor = None
         else:
-            logger.debug("Found matching device %s", reserved_device)
+            logger.debug("Found matching device %s", reserved_gpu_device)
             success = True
```
Comment (richardliaw) [inline python/ray/util/sgd/torch/distributed_torch_runner.py:247]: can you please revert this change?

**Comment by richardliaw** [comment]
> Hi @richardliaw, could you help to review on this? Thanks a lot. And also, I have a question whether should we support some resource init function such as the following?
> 
> ```python
> def resource_requirement(rank_id):
>     #  specify the resource requirement for the given rank worker
>     return resources
> ```
> 
> I think this is helpful because of user can set the specified resource requirement for each training worker.

What would be your use case for this?

File: python/ray/util/sgd/torch/distributed_torch_runner.py (hunk 1 above)
Comment (ConeyLiu) [reply python/ray/util/sgd/torch/distributed_torch_runner.py:247]: Now the dummy actor reserve both CPU and GPU resources. The dummy actors will be reserved many of CPU resources if GPU reserving failed multiple times. Does this expect?

**Comment by ConeyLiu** [comment]
> What would be your use case for this?

For clusters of heterogeneous resources, I want to schedule tasks to a given node.

File: python/ray/util/sgd/torch/distributed_torch_runner.py (hunk 1 above)
Comment (ConeyLiu) [reply python/ray/util/sgd/torch/distributed_torch_runner.py]: reverted

File: python/ray/util/sgd/torch/distributed_torch_runner.py [hunk 2]
```diff
@@ -271,33 +289,35 @@ def __init__(self, *args, num_cpus=None, num_gpus=None, **kwargs):
             assert num_gpus == 1, "Does not support multi-gpu workers"
 
         if not self.is_actor() and num_gpus > 0:
-            self._try_reserve_and_set_cuda()
+            self._try_reserve_resources_and_set_cuda()
 
         super(LocalDistributedRunner, self).__init__(*args, **kwargs)
 
-    def _try_reserve_and_set_cuda(self):
+    def _try_reserve_resources_and_set_cuda(self, num_cpus, num_gpus):
```
Comment (richardliaw) [inline python/ray/util/sgd/torch/distributed_torch_runner.py:296]: I think there is a typo here

File: python/ray/util/sgd/torch/distributed_torch_runner.py [hunk 3]
```diff
@@ -271,33 +289,35 @@ def __init__(self, *args, num_cpus=None, num_gpus=None, **kwargs):
             assert num_gpus == 1, "Does not support multi-gpu workers"
 
         if not self.is_actor() and num_gpus > 0:
-            self._try_reserve_and_set_cuda()
+            self._try_reserve_resources_and_set_cuda()
```
Comment (richardliaw) [inline python/ray/util/sgd/torch/distributed_torch_runner.py:292]: this signature doesn't match the below defined signature

**Comment by richardliaw** [comment]
Sorry, this slipped my mind. Can you fix the comments and add a small test for CPU?

Thanks!

**Comment by ConeyLiu** [comment]
Updated

File: python/ray/util/sgd/tests/test_torch_runner.py [hunk 4]
```diff
@@ -203,54 +203,96 @@ def _testWithInitialized(self, init_mock):
...
+        LocalDistributedRunner._try_reserve_and_set_resources(
+            mock_runner, 4, 0)
+
+        available_devices = set(range(0, 10))
+
+        self.assertTrue(mock_runner._set_cpu_devices.called)
+        set_devices = mock_runner._set_cpu_devices.call_args[0][0]
+        self.assertEquals(len(set_devices), 4)
+        self.assertTrue(set(set_devices).issubset(available_devices))
+
```
Comment (richardliaw) [inline python/ray/util/sgd/tests/test_torch_runner.py:290]: OK instead of putting code into the runner, how about just checking the ray global state? `ray.available_resources()["CPU"]` should indicate that the remaining CPUs is less than the total cluster_resources.

File: python/ray/util/sgd/tests/test_torch_runner.py [hunk 5]
```diff
@@ -203,54 +203,96 @@ def _testWithInitialized(self, init_mock):
...
+        os.environ["CUDA_VISIBLE_DEVICES"] = "2,3"
+        with patch("torch.cuda.is_initialized") as init_mock:
+            init_mock.return_value = False
+            self._testReserveCUDAResource(init_mock, 0)
+
+    def test2VisibleNotInitializedAndReserveCPUResource(self):
+        os.environ["CUDA_VISIBLE_DEVICES"] = "2,3"
+        with patch("torch.cuda.is_initialized") as init_mock:
+            init_mock.return_value = False
+            self._testReserveCUDAResource(init_mock, 2)
```
Comment (richardliaw) [inline python/ray/util/sgd/tests/test_torch_runner.py:266]: I don't think we need this test, it seems redundant

File: python/ray/util/sgd/tests/test_torch_runner.py [hunk 6]
```diff
@@ -203,54 +203,96 @@ def _testWithInitialized(self, init_mock):
...
             init_mock.return_value = True
-            self._testWithInitialized(init_mock)
+            self._testReserveCUDAResource(init_mock, 0)
 
-    def test2VisibleNotInitialized(self):
+    def test1VisibleWithInitializedAndReserveCPUResource(self):
+        os.environ["CUDA_VISIBLE_DEVICES"] = "0"
+        with patch("torch.cuda.is_initialized") as init_mock:
+            init_mock.return_value = True
+            self._testReserveCUDAResource(init_mock, 2)
```
Comment (richardliaw) [inline python/ray/util/sgd/tests/test_torch_runner.py:230]: Also don't think we need this test, it seems redundant

File: python/ray/util/sgd/torch/distributed_torch_runner.py [hunk 7]
```diff
@@ -283,41 +295,55 @@ def __init__(self, *args, num_cpus=None, num_gpus=None, **kwargs):
...
         super(LocalDistributedRunner, self).__init__(*args, **kwargs)
 
-    def _try_reserve_resources_and_set_cuda(self, num_cpus, num_gpus):
-        visible_devices = os.environ.get("CUDA_VISIBLE_DEVICES")
-        reserved_gpu_device = reserve_resources(num_cpus, num_gpus)
+    def _try_reserve_and_set_resources(self, num_cpus, num_gpus):
+        visible_cuda_devices = os.environ.get("CUDA_VISIBLE_DEVICES")
+        reserved_cpu_devices, reserved_cuda_device = reserve_resources(
+            num_cpus, num_gpus)
+        self._set_cpu_devices(reserved_cpu_devices)
```
Comment (richardliaw) [inline python/ray/util/sgd/torch/distributed_torch_runner.py:319]: I don't think we need to do this, see comments on the test

File: python/ray/util/sgd/torch/distributed_torch_runner.py [hunk 8]
```diff
@@ -283,41 +295,55 @@ def __init__(self, *args, num_cpus=None, num_gpus=None, **kwargs):
...
             self._set_cuda_device(str(scoped_index))
         else:
             # Once cuda is initialized, torch.device ignores the os.env
             # so we have to set the right actual device.
-            self._set_cuda_device(reserved_gpu_device)
+            self._set_cuda_device(reserved_cuda_device)
+
+    def _set_cpu_devices(self, device_list):
+        # for test
+        self.local_cpu_device = device_list
```
Comment (richardliaw) [inline python/ray/util/sgd/torch/distributed_torch_runner.py:346]: don't think we need to do this, see comments on the test

File: python/ray/util/sgd/torch/distributed_torch_runner.py [hunk 9]
```diff
@@ -220,25 +247,25 @@ def reserve_cuda_device(retries=20):
...
     unused_actors = []
 
     success = False
     for _ in range(retries):
-        if _dummy_actor is None:
-            _dummy_actor = ray.remote(
-                num_gpus=1,
+        if _dummy_cuda_actor is None:
+            _dummy_cuda_actor = ray.remote(
+                num_cpus=0, num_gpus=num_gpus,
```
Comment (ConeyLiu) [inline python/ray/util/sgd/torch/distributed_torch_runner.py:257]: here we specify the `num_cpus=0` because CPU reservation should be done by the `_dummy_cpu_actor `

File: python/ray/util/sgd/tests/test_torch_runner.py (hunk 4 above)
Comment (ConeyLiu) [reply python/ray/util/sgd/tests/test_torch_runner.py]: updated

File: python/ray/util/sgd/tests/test_torch_runner.py (hunk 6 above)
Comment (ConeyLiu) [reply python/ray/util/sgd/tests/test_torch_runner.py]: removed

File: python/ray/util/sgd/tests/test_torch_runner.py (hunk 5 above)
Comment (ConeyLiu) [reply python/ray/util/sgd/tests/test_torch_runner.py]: removed

File: python/ray/util/sgd/torch/distributed_torch_runner.py (hunk 7 above)
Comment (ConeyLiu) [reply python/ray/util/sgd/torch/distributed_torch_runner.py]: removed

File: python/ray/util/sgd/torch/distributed_torch_runner.py (hunk 8 above)
Comment (ConeyLiu) [reply python/ray/util/sgd/torch/distributed_torch_runner.py]: removed

File: python/ray/util/sgd/torch/distributed_torch_runner.py (hunk 9 above)
Comment (richardliaw) [reply python/ray/util/sgd/torch/distributed_torch_runner.py:257]: sgtm

**Review by richardliaw — APPROVED** [review]

**Comment by richardliaw** [comment]
looks good to me; please ping me when tests pass!

**Comment by ConeyLiu** [comment]
thanks @richardliaw
